Add console entry point script for binary download

paulmueller opened the pull request:
Please do not merge yet. There are a few points open for discussion.

Here is a proposal for an entry point script for binary download (related to issue #32)

The new entry point script can be seen in action in this appveyor build (line 246):
https://ci.appveyor.com/project/paulmueller/ShapeOut/build/1.0.688/job/0mo8iu4xxs6on968#L246

There are a few things that need to be clarified before merging:
1. Are you fine with a separate `scripts.py` module or would you like to have this functionality someplace else?
2. To allow the option to download all available binaries, I need a method that gets all plugins with binary files. One possibility is to check all submodules in `plugins` for the `download` method. Is there a better way? 
3. I would like to add `*args` and `**kwargs` to the `download` methods for ffmpeg, avbin, and freeimage  to be able force-download the binary files into the `recources` directory (see #32). Would that be an acceptable change?
4. Could you please point me to the relevant parts of the documentation that require an update due to this PR?

almarklein commented:
Thanks for this!

> Are you fine with a separate scripts.py module or would you like to have this functionality someplace else?

I like `__main__.py` for this, since it also allows invoking the entry point through `python -m imageio ...`. E.g. for cases where the package is not installed (e.g. dev env), or to explicitly specify the Python interpreter.

> To allow the option to download all available binaries, I need a method that gets all plugins with binary files. One possibility is to check all submodules in plugins for the download method. Is there a better way?

Mmm, I think it should be enough to specify a list of plugin names at the top of the script. Checking plugins for a download method might break when a plugin implements a download method (for whatever reason) that does not do what the script expects it to do. There's not a lot of plugins that need downloads, so it should not be hard to maintain this. If needed, we could look at automating this, but then use a more specific function name.

> I would like to add `*args` and `**kwargs` to the download methods for ffmpeg, avbin, and freeimage to be able force-download the binary files into the recources directory (see #32). Would that be an acceptable change?

I think I'd prefer a keyword argument (e.g. `def download(force_download=False)`) and then pass that to where its needed.

> Could you please point me to the relevant parts of the documentation that require an update due to this PR?

* I think we need a new page, i.e. a subling of https://github.com/imageio/imageio/blob/master/docs/envvariables.rst
* Add it in the index here: https://github.com/imageio/imageio/blob/master/docs/sec_reference.rst, I think right before the section on environment variables.
* The docs in the ffmpeg and freeimage plugin should mention both the CLI  and `imageio.plugins.xx.download()` as a way to download the binaries. They still seem to refer to an automated download, which is no longer the case.

paulmueller commented:
Ok, I addressed all your points except for the docs.
I had to edit the [ffmpeg.get_exe](https://github.com/imageio/imageio/pull/270/files#diff-d72802e02eea93d5f9fa4aa279eeaef2L61) method so that the imageio binaries get highest priority (this is especially important when shipping a frozen application). Please have a look and let me know if that is ok.

paulmueller commented:
Update: Creating a distribution with PyInstaller works.
Here is the pull request with the hook script:
https://github.com/pyinstaller/pyinstaller/pull/2696

almarklein commented on imageio/core/util.py:
```diff
@@ -495,8 +497,10 @@ def resource_dirs():
+    # Resource dir baked in the package.
+    # (The directory returned by `pkg_resources.resource_filename`
+    # also works with eggs.)
+    dirs.append(pkg_resources.resource_filename("imageio", "resources"))
```
`pkg_resources` comes with setuptools, right? Can we please make it optional? I want to avoid cases like [this](https://stackoverflow.com/questions/7446187/no-module-named-pkg-resources).

E..g.
```py
try:
   import pkg_resources
except ImportError:
    pkg_resources = None

...

if pkg_resources:
    use pkg_resources
else:
   use the old line
```

almarklein commented on imageio/plugins/ffmpeg.py:
```diff
@@ -91,33 +115,39 @@ def get_exe():
             with open(os.devnull, "w") as null:
                 sp.check_call([exe, "-version"], stdout=null, stderr=sp.STDOUT)
-                return exe
+                exes.append(exe)
```
I guess this should be `insert`?

Maybe its easiest to just append to `exes`, in the order of priority, and add `'ffmpeg'` last. 

I think it would be ok to give highest priority to the one provided by imageio. But it would be good to have a way to remove the binary to have more control, e.g. in case its corrupt. Can you add that to the CLI?

paulmueller replied:
Sure, this can be `insert` to give it priority over the  system's ffmpeg.

If we want to append the exes in order of priority, then we would have to refactor the case in lines 122+.

Sure, I can add a `--remove` switch to the entry point.

paulmueller replied:
see https://github.com/imageio/imageio/pull/270/commits/40655463b0c215cd5c4a9d114c2795d3dea0e5e1: I refactored out `resource_package_dir` which is also used in `__main__`.

paulmueller replied:
see https://github.com/imageio/imageio/pull/270/commits/956955ef187aeed6fe3f6839b41c9324e4c105cd: I added the `imageio_remove_bin` entry point.

paulmueller commented:
Please have a look. If you have nothing to add, this could be merged now.

almarklein commented on imageio/__main__.py:
```diff
@@ -0,0 +1,137 @@
+    
+    for plg in plugin_names:
+        msg = "Plugin {} is not registered for binary download!".format(plg)
+        assert plg in PLUGINS_WITH_BINARIES, msg
```
Can you make this a test + raise, assert statements are removed from optimized bytecode.

almarklein commented on imageio/__main__.py:
```diff
@@ -0,0 +1,137 @@
+
+    for plg in plugin_names:
+        msg = "Plugin {} is not registered for binary download!".format(plg)
+        assert plg in PLUGINS_WITH_BINARIES, msg
```
Dito

almarklein commented on imageio/__main__.py:
```diff
@@ -0,0 +1,137 @@
+        # plugin name is in subdirectories
+        for rsub in os.listdir(rd):
+            if rsub in plugin_names:
+                shutil.rmtree(op.join(rd, rsub))
```
Maybe have a try-except around this? When the lib is still loaded, a remove may fail, preventing the removal for other dirs. Perhaps raise error at the end when any of the deletions failed so that the process returns -1.

almarklein commented on imageio/plugins/ffmpeg.py:
```diff
@@ -66,22 +82,16 @@ def get_exe():
     
     # Check if ffmpeg is installed in Python environment
     # (e.g. via conda install ffmpeg -c conda-forge)
-    exe = None
```
I'm not yet happy with this part :) E.g. exes is always nonempty, but at the end we test whether it is.

I think it would be better to forget about that list and just try each option in order of priority, returning the value if it was found to be correct.

* If `IMAGEIO_FFMPEG_EXE` is set, use that (even without testing whether it is a valid path).
* Try our own version using `get_remote_file()`. (if `NeedDownloadError` is raised, just continue to next step)
* Try from conda package
* Try `ffmpeg` command
* Nothing found, raise `NeedDownloadError` with nice message.

almarklein commented on imageio/plugins/ffmpeg.py:
```diff
@@ -143,11 +158,15 @@ class FfmpegFormat(Format):
     efficient.
     
+    The ffmpeg plugin requires an `ffmpeg` binary. If this binary is
+    not available on the system, it can be downloaded by either
```
Would be nice to document the places where imageio tries to load ffmpeg from.

almarklein replied:
Well, it sort of is already, but maybe making the priority more explicit.

almarklein commented:
I apologize for the late reply. I was away on holiday and then a conference, and it disappeared from my radar (i.e. I probably accidentally marked the notification email as unread).

This is looking really good! I made a few small comments, and a slightly larger one regarding ffmpeg's download sequence.

A test would be great, also to prevent the coverage to drop too much. It might be a bit hard to fully test it fully, but some unit tests that cover most of the new code would be great.

almarklein commented on imageio/__main__.py:
```diff
@@ -0,0 +1,137 @@
+
+
+if __name__ == "__main__":
+    download_bin_main()
```
For testing and dev, it might be nice to support `python -m imageio download_bin` and `python -m imageio remove_bin`. As it is now, I think one could call `python -m imageio` and it would default to `download_bin_main()`. In short something along these lines should work:
```py
if __name__ == "__main__":  # technically always True for a __main__.py
    if len(sys.argv) > 1 and sys.argv[1] == 'download_bin':
        download_bin_main()
    elif len(sys.argv) > 1 and sys.argv[1] == 'remove_bin':
        remove_bin_main()
    else:
        raise RuntimeError('Invalid use of the imageio CLI')
```

paulmueller commented:
I covered everything you mentioned and added a few tests.

If you are ok with it, I would not like to implement compatibility for binary download in Python 2.6.
The problem is similar to this one: https://stackoverflow.com/questions/19668395/str-format-for-python-2-6-gives-error-where-2-7-does-not

I think it makes the code unreadable if I added zeros at every use of "format" and I think no one should be using Python2.6 anyway. Do you agree?

see https://github.com/paulmueller/imageio/blob/406a2166a627e81a0b78ceb66be0d7d1e79332ed/tests/test_bin.py

almarklein commented:
> If you are ok with it, I would not like to implement compatibility for binary download in Python 2.6.

Fair enough.

> I think no one should be using Python2.6 anyway. Do you agree?

Hell yeah :)

This is looking good. Is it ready from your end?

paulmueller commented:
![image](https://user-images.githubusercontent.com/2853022/30312318-386cc1fc-9799-11e7-84e5-3664b9cd8a21.png)

almarklein commented:
Thank you!
